lib/api/units/validateBody.js

```diff
@@ -0,0 +1,233 @@
+const jph = require('json-parse-helpfulerror');
+const mediaTyper = require('media-typer');
+const contentTypeUtils = require('content-type');
+
+const { TextDiff } = require('../../../lib/validators/text-diff');
+const { JsonExample } = require('../../../lib/validators/json-example');
+const { JsonSchema } = require('../../../lib/validators/json-schema');
+
+function isPlainText(mediaType) {
+  return mediaType.type === 'text' && mediaType.subtype === 'plain';
+}
+
+function isJson(mediaType) {
+  return mediaType.type === 'application' && mediaType.subtype === 'json';
+}
+
+function isJsonWeak(mediaType) {
+  return isJson(mediaType) || mediaType.suffix === 'json';
+}
+
+/**
+ * Parses a given content-type header into media type.
+ * @param {string} contentType
+ * @returns {[Error, MediaType]}
+ */
+function parseContentType(contentType) {
+  try {
+    const { type } = contentTypeUtils.parse(`${contentType}`);
+    const mediaType = mediaTyper.parse(type);
+    return [null, mediaType];
+  } catch (error) {
+    return [error, null];
+  }
+}
+
+/**
+ * Determines if a given 'Content-Type' header contains JSON.
+ * @param {string} contentType
+ * @returns {boolean}
+ */
+function isJsonContentType(contentType) {
+  // No need to explicitly check for "contentType" existence.
+  // In case given content type is invalid, "parseContentType"
+  // will return an error, which would be coerced to "false".
+  const [error, mediaType] = parseContentType(contentType);
+
+  // Silence an error on purporse because contentType
+  // may contain any kind of rubbish.
+  return error ? false : isJsonWeak(mediaType);
+}
+
+/**
+ * Returns a tuple of error and body media type based
+ * on the given body and normalized headers.
+ * @param {string} body
+ * @param {Object} headers
+ * @returns {[error, bodyType]}
+ */
+function getBodyType(body, headers) {
+  const contentType = headers && headers['content-type'];
+  const hasJsonContentType = isJsonContentType(contentType);
+
+  try {
+    jph.parse(body);
+    const [mediaTypeError, bodyMediaType] = parseContentType(
+      hasJsonContentType ? contentType : 'application/json'
+    );
+
+    if (mediaTypeError) {
+      throw new Error(mediaTypeError);
+    }
+
+    return [null, bodyMediaType];
+  } catch (parsingError) {
+    const fallbackMediaType = mediaTyper.parse('text/plain');
+    const error = hasJsonContentType
+      ? {
+          /**
+           * @TODO The same message for real/expected
+           * body assertion. Real/expected must be reflected
+           * in the error message.
+           */
+          message: `\
+Real body 'Content-Type' header is '${contentType}' \
+but body is not a parseable JSON:
+${parsingError.message}\
```

Comment:
No, definitely not. Previously, Gavel operated on "raw" content type, which required to parse it multiple times on some occasions. To prevent that, I introduced the storage of content-type as parsed throughout the validation unit logic.

I completely agree, I've also voiced my concern in other comment that this `throw` statement on media type parsing error doesn't belong there. It's questionable if it even should be handled.

Reply:
I wonder what defines `content type NOT parseable`?

Would it be legible for validation if you explicitly specify you expect a certain content type, but the real entity contains a different, or no content-type at all? I'm examining how Gavel works, and it uses weak comparison for headers, sometimes checking only the header names (without values), sometimes behaving like [`containsAllKeys`](https://www.chaijs.com/api/assert/#method_containsallkeys) from `chai`. I can see this leading to false positives.

Reply:
Yeah, reading the table again, I think perhaps the fact that content type isn't parseable is not even worth the warning in fact?

When I think about it, basically the user wants to know whether the real and expected HTTP messages fit, they probably don't care so much about whether the content type is parseable or not. Gavel doesn't validate the inputs according to HTTP specs, it just silently does a best guess and provides the verdict and the diff.

So I'd change "warns about invalid content type, returns application/json (for the purpose of validation)" to "returns application/json (for the purpose of validation)" in case the body is parseable as JSON, but the content type is rubbish. That way the exception gets silenced, as you suggest.

Reply:
One thing is validation of the body, where we can infer certain behavior based on the value of the content type header, or to use whatever effective media type we think is the best. In that case I think it's only our internal implementation detail that we're trying to parse the media type and that it can fail if the value is rubbish. I think we should silence the exception and return whatever we infer from the situation.

Another thing is validation of the headers. As far as I recall Gavel validates only presence of headers (i.e. keys), except of content negotiation headers (content type, accept, etc.), where it also looks at the values. When validating the content type header in the context of headers (not body) validation, we don't even need to parse it. It's plain 1:1 comparison. If it's specified it should be `application/json`, then `application/json; charset=utf-8` fails the validation. There is only one corner case I can see, that is `application/json; charset=utf-8` vs `application/json; charset=UTF-8`, which I think Gavel should consider the same - https://github.com/apiaryio/gavel.js/issues/131, but that is something we can fix as the next step. And still, it can be carried out gracefully (pseudocode):

```
function normalizeContentTypeValue(contentType) { // for the purpose of headers validation
  try {
    const mediaType = parseMediaType(contentType)
    if (mediaType.params.charset) {
      lowercase the charset, format the media type back to string, return
    }
    return contentType;
  } catch (e) {
    return contentType;
  }
}
```

Reply:
> Or maybe we can just lowercase the whole header value?

Absolutely. As you've said, it's a plain string comparison, and such should always be case-insensitive.